### PR TITLE
Object Oriented JS (frogger)

### PR DESCRIPTION
**Pull request by MosQuitO404**
### Object Oriented JS

[Demo](https://mosquito404.github.io/frontend-nanodegree-arcade-game/) |
[Code base](https://github.com/MosQuitO404/frontend-nanodegree-arcade-game/tree/master/js)

The code is submitted in a dedicated feature branch.

Only code files are submitted.

Please, review.

### CODE REVIEW

**Comment by MosQuitO404** [comment]
I have completed self-checks and fixed code accordingly

File: submissions/MosQuitO404/OOPJS_frogger/app.js [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    return Math.floor(Math.random() * (maxSpeed - minSpeed)) + minSpeed; 
+}
+
+Enemy.prototype.checkCollisions = function() {
+    if (player.y < this.y + player.collisionAreaHeight && 
+        player.y + player.collisionAreaHeight > this.y &&
+        player.x < this.x + player.collisionAreaWidth && 
+        player.x + player.collisionAreaWidth > this.x) {
+            player.x = player.columnWidth * 2; 
+            player.y = player.rowHeight * 5;
```
Comment (OleksiyRudenko) [inline submissions/MosQuitO404/OOPJS_frogger/app.js:51]: See item 4.iii of the requirements

File: submissions/MosQuitO404/OOPJS_frogger/app.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    this.checkCollisions();
+};
+
+// Draw the enemy on the screen, required method for game
+Enemy.prototype.render = function() {
+    ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+Enemy.prototype.positionEnemyMaker = function() {
+    const allowedPositions = [47, 129, 211];
```
Comment (OleksiyRudenko) [inline submissions/MosQuitO404/OOPJS_frogger/app.js:37]: Imagine you are tasked to add two more rows. Please optimize this. 
Rows seem to be of the same height. Something is off in your calculations. Let computer do calculations.

File: submissions/MosQuitO404/OOPJS_frogger/app.js [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    this.sprite = 'images/enemy-bug.png';
+};
+
+// Update the enemy's position, required method for game
+// Parameter: dt, a time delta between ticks
+Enemy.prototype.update = function(dt) {
+    // You should multiply any movement by the dt parameter
+    // which will ensure the game runs at the same speed for
+    // all computers.
+    if (this.x < 500) {
```
Comment (OleksiyRudenko) [inline submissions/MosQuitO404/OOPJS_frogger/app.js:20]: 50, -95 and other numbers seem mysterious.
See section 3 of the requirements.

File: submissions/MosQuitO404/OOPJS_frogger/app.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+            player.y = player.rowHeight * 5;
+    }
+};
+
+// Now write your own player class
+// This class requires an update(), render() and
+// a handleInput() method.
+const Player = function() {
+    this.rowHeight = 75;
+    this.columnWidth = 101;
```
Comment (OleksiyRudenko) [inline submissions/MosQuitO404/OOPJS_frogger/app.js:60]: These do not seem to be properties of a player. Player doesn't have rows or columns.

File: submissions/MosQuitO404/OOPJS_frogger/app.js [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        this.winReset();   
+    }
+};
+
+// Now instantiate your objects.
+// Place all enemy objects in an array called allEnemies
+// Place the player object in a variable called player
+let allEnemies = [];
+function createEnemies(minEnemies, maxEnemies) {
+    for (let i = 0; i < Math.floor(Math.random() * (maxEnemies - minEnemies + 1)) + minEnemies; i++) {
```
Comment (OleksiyRudenko) [inline submissions/MosQuitO404/OOPJS_frogger/app.js:112]: `i < Math.floor(Math.random() * (maxEnemies - minEnemies + 1)) + minEnemies` runs at every loop iteration.
Is it what you actualy want?

File: submissions/MosQuitO404/OOPJS_frogger/app.js [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+Player.prototype.winReset = function() {
+    if (this.y < 0) {
+        setTimeout(() => {
+            alert('Great! You Win!');
+            this.x = this.columnWidth * 2; 
+            this.y = this.rowHeight * 5;
+            allEnemies = [];
```
Comment (OleksiyRudenko) [inline submissions/MosQuitO404/OOPJS_frogger/app.js:82]: See item 4.iii of the requirements.

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@MosQuitO404 good start.
Needs improvement.
Comments below refer to these [requirements](https://kottans.org/documentation/docs/doc/code-review/#oo-js-frogger)

File: submissions/MosQuitO404/OOPJS_frogger/app.js (hunk 1 above)
Comment (MosQuitO404) [reply submissions/MosQuitO404/OOPJS_frogger/app.js:51]: Ok, done! Thanks

File: submissions/MosQuitO404/OOPJS_frogger/app.js (hunk 2 above)
Comment (MosQuitO404) [reply submissions/MosQuitO404/OOPJS_frogger/app.js:37]: Resolved

File: submissions/MosQuitO404/OOPJS_frogger/app.js (hunk 3 above)
Comment (MosQuitO404) [reply submissions/MosQuitO404/OOPJS_frogger/app.js:20]: Ok, resolved. It's canvas size :)

File: submissions/MosQuitO404/OOPJS_frogger/app.js (hunk 4 above)
Comment (MosQuitO404) [reply submissions/MosQuitO404/OOPJS_frogger/app.js:60]: Ok! Done

File: submissions/MosQuitO404/OOPJS_frogger/app.js (hunk 6 above)
Comment (MosQuitO404) [reply submissions/MosQuitO404/OOPJS_frogger/app.js:82]: Ok! Done

File: submissions/MosQuitO404/OOPJS_frogger/app.js (hunk 5 above)
Comment (MosQuitO404) [reply submissions/MosQuitO404/OOPJS_frogger/app.js:112]: Yes. Every start new game this iteration randomly generate from 3 to 5 enemies.

**Comment by MosQuitO404** [comment]
@A-Ostrovnyy Hi! Please mark label for Stage 0.1

File: submissions/MosQuitO404/OOPJS_frogger/app.js (hunk 5 above)
Comment (OleksiyRudenko) [reply submissions/MosQuitO404/OOPJS_frogger/app.js:115]: But the quoted fragment is executed not only for every new game but literally at every `for` loop iteration. `i` gets compared to a different number at every `i++` change.
What sense does it make?

Comment (MosQuitO404) [reply submissions/MosQuitO404/OOPJS_frogger/app.js]: @OleksiyRudenko Thanks! You're right. Please see resolve in code. So maybe is this right solution?

**Review by OleksiyRudenko — APPROVED** [review]
@MosQuitO404 great job!

**Comment by MosQuitO404** [comment]
> @MosQuitO404 great job!

@OleksiyRudenko Thank you very much! The best motivation for me.

**Pull request merged**